cmd/npmExecuteScripts.go

```diff
@@ -0,0 +1,198 @@
+	return doublestar.Glob(pattern)
+}
+
+func (u *npmExecuteScriptsUtilsBundle) getwd() (dir string, err error) {
+	return os.Getwd()
+}
+
+func (u *npmExecuteScriptsUtilsBundle) chdir(dir string) error {
+	return os.Chdir(dir)
+}
+
+func (u *npmExecuteScriptsUtilsBundle) getExecRunner() execRunner {
+	if u.execRunner == nil {
+		u.execRunner = &command.Command{}
+		u.execRunner.Stdout(log.Entry().Writer())
+		u.execRunner.Stderr(log.Entry().Writer())
+	}
+	return u.execRunner
+}
+
+func npmExecuteScripts(config npmExecuteScriptsOptions, telemetryData *telemetry.CustomData) {
+	utils := npmExecuteScriptsUtilsBundle{}
+
+	err := runNpmExecuteScripts(&utils, &config)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+func runNpmExecuteScripts(utils npmExecuteScriptsUtilsInterface, options *npmExecuteScriptsOptions) error {
+	execRunner := utils.getExecRunner()
+	packageJSONFiles, err := findPackageJSONFiles(utils)
+	if err != nil {
+		return err
+	}
+
+	oldWorkingDirectory, err := utils.getwd()
+
+	for _, file := range packageJSONFiles {
+		dir := path.Dir(file)
+		err = utils.chdir(dir)
+		if err != nil {
+			return err
+		}
+
+		// set in each directory to respect existing config in rc files
+		err = setNpmRegistries(options, execRunner)
+		if err != nil {
+			return err
+		}
+
+		packageLockExists, yarnLockExists, err := checkIfLockFilesExist(utils)
+		if err != nil {
+			return err
+		}
+		if options.Install {
+			err = installDependencies(dir, packageLockExists, yarnLockExists, execRunner)
+			if err != nil {
+				return err
+			}
+		}
+
+		for _, v := range options.RunScripts {
+			log.Entry().WithField("WorkingDirectory", dir).Info("run-script " + v)
+			err = execRunner.RunExecutable("npm", "run-script", v, "--if-present")
+			if err != nil {
+				return err
+			}
+		}
+		err = utils.chdir(oldWorkingDirectory)
+		if err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+func setNpmRegistries(options *npmExecuteScriptsOptions, execRunner execRunner) error {
+	var environment []string
+	configurableRegistries := []string{"registry", "@sap:registry"}
+	var buffer bytes.Buffer
+	execRunner.Stdout(&buffer)
+	for _, registry := range configurableRegistries {
+		err := execRunner.RunExecutable("npm", "config", "get", registry)
+		if err != nil {
+			return err
+		}
+		preConfiguredRegistry := buffer.String()
+
+		if preConfiguredRegistry == "undefined" {
+			if registry == "registry" {
+				log.Entry().Info("npm registry " + registry + " was not configured, setting it to " + options.DefaultNpmRegistry)
+				environment = append(environment, "npm_config_"+registry+"="+options.DefaultNpmRegistry)
+			}
+
+			if registry == "@sap:registry" {
+				log.Entry().Info("npm registry " + registry + " was not configured, setting it to " + options.SapNpmRegistry)
+				environment = append(environment, "npm_config_"+registry+"="+options.SapNpmRegistry)
+			}
+		}
+	}
+	defer execRunner.Stdout(log.Entry().Writer())
```

Comment:
This line needs to go directly onto line 115, after changing the Stdout. It is then "deferred" to when the scope is left, i.e. when the function returns in any way. If you return early in line 118, this code is never reached.